### PR TITLE
Feature/1781 Twilio is integrated with app for event triggered notifications

### LINKED ISSUE #1781: Twilio is integrated with app for event triggered notifications
Part of epic #1017, dependent on #1780

AS a maintainer (and/or possibly stakeholder),
I WANT there to be a place on platform (either under the hood or on the frontend) to add a Twilio API key,
SO THAT the application will be able to trigger Twilio SMS notifications.

### How it works now
It doesn't! Twilio is new territory for Ruby For Good applications, including this project.

### How it should work
There are a couple ways this could work, depending on the outcome of #1780. It looks likely that there will need to be a place on platform (on the frontend) for stakeholders or maintainers to enter a tenant specific Twilio API key.

This could either go: 
- in prod under "Edit Organization Details", only accessible by admins
- in prod All CASA Admin editable per organization, only accessible by maintainers

This could alternatively be hard coded into the backend (especially if the app will be relying on one API key vs numerous), but probably easier to maintain if we create a place for it on the frontend.

### PR DESCRIPTION
**Pull request by harsohailB**
### What github issue is this PR for, if any?

Resolves #1781 

### What changed, and why?

* Added `Twilio Account SID`, `Twilio API Key SID`, `Twilio API Key Secret`, and `Twilio Phone Number` under the "Edit Organization" tab

These are given by Twilio in their console as follows:

Screen Shot 2022-05-09 at 12 22 56 PM

Screen Shot 2022-05-09 at 5 19 10 PM

A POST request like the following is sent to send an SMS which requires these three parameters:

* Changed `CasaOrg` model to store the three attributes

```
curl -X POST https://api.twilio.com/2010-04-01/Accounts/${TWILIO_ACCOUNT_SID}/Messages.json \
                --data-urlencode "Body=Hello from Twilio" \
                --data-urlencode "From=${TWILIO_PHONE_NUMBER}" \
                --data-urlencode "To=${RECIPIENT NUMBER}" \
                -u ${TWILIO_API_KEY_SID}:${TWILIO_API_KEY_SECRET}
```

### How will this affect user permissions?
- Volunteer permissions: n/a
- Supervisor permissions: n/a
- Admin permissions: n/a

### How is this tested? (please write tests!) 💖💪

* Wrote view tests for the "Edit Organization" page which now contains Twilio inputs
* Wrote requests tests to verify update requests for a `CasaOrg` (invalid phone number input not accepted)

### Screenshots please :)

Screen Shot 2022-05-10 at 11 31 09 AM

### CODE REVIEW

File: app/validators/casa_org_validator.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  VALID_TWILIO_PHONE_NUMBER_LENGTH = 12
+  VALID_TWILIO_COUNTRY_CODE = "+1"
+
+  def validate(record)
+    valid_twilio_phone_number(record.twilio_phone_number, record)
+  end
+
+  private
+
+  def valid_twilio_phone_number(number, record)
```
Comment (xihai01) [inline app/validators/casa_org_validator.rb:11]: @harsohailB Can this method be DRY'd up? It seems awfully similar to the phone number validation method for validators/user_validator.rb

File: db/migrate/20220508232038_add_columns_to_casa_orgs.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class AddColumnsToCasaOrgs < ActiveRecord::Migration[7.0]
+  def change
+    add_column :casa_orgs, :twilio_account_sid, :string
```
Comment (xihai01) [inline db/migrate/20220508232038_add_columns_to_casa_orgs.rb:3]: @harsohailB I'm wondering, but should these API credentials be hashed before storing them in the db? https://security.stackexchange.com/questions/180345/do-i-need-to-hash-or-encrypt-api-keys-before-storing-them-in-a-database

Maybe @compwron can give some suggestions.

File: spec/requests/casa_org_spec.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 RSpec.describe "CasaOrg", type: :request do
   let(:casa_org) { build(:casa_org) }
-  let(:valid_attributes) { {name: "name", display_name: "display_name", address: "address"} }
+  let(:valid_attributes) { {name: "name", display_name: "display_name", address: "address", twilio_account_sid: "fyxpz5naqnir3ftopvxuzev6ir48xb4jmc", twilio_auth_token: "gukogx4k99885clw6j7ucd62hgzj5w4p", twilio_phone_number: "+12223334444"} }
```
Comment (xihai01) [inline spec/requests/casa_org_spec.rb:5]: @harsohailB just making sure, but these API credentials are fake right?

**Review by xihai01 — APPROVED** [review]
awesome stuff! 😄 Left a few comments for you. I also have a suggestion. According to the Twilio docs, they recommend using API keys since they aren't attached to the main account and can be easily issued and revoked: https://www.twilio.com/docs/usage/rest-api-best-practices#control-access-to-your-accounts

File: db/migrate/20220508232038_add_columns_to_casa_orgs.rb (hunk 2 above)
Comment (compwron) [reply db/migrate/20220508232038_add_columns_to_casa_orgs.rb:3]: In general yes. Use https://guides.rubyonrails.org/active_record_encryption.html

**Comment by compwron** [comment]
Yes, API keys- and stored as env variables

File: spec/requests/casa_org_spec.rb (hunk 3 above)
Comment (harsohailB) [reply spec/requests/casa_org_spec.rb:5]: Yes they are fake

File: app/validators/casa_org_validator.rb (hunk 1 above)
Comment (harsohailB) [reply app/validators/casa_org_validator.rb:11]: Will do!

File: db/migrate/20220508232038_add_columns_to_casa_orgs.rb (hunk 2 above)
Comment (harsohailB) [reply db/migrate/20220508232038_add_columns_to_casa_orgs.rb:3]: I will look into that and hash before storing!

File: app/validators/casa_org_validator.rb (hunk 1 above)
Comment (harsohailB) [reply app/validators/casa_org_validator.rb:10]: Abstracted out phone number validity logic into a helper named `PhoneNumberHelper`. Will create a ticket to point our previously worked on `UserValidator` and importers to this as well for code reuse.

**Comment by harsohailB** [comment]
> awesome stuff! 😄 Left a few comments for you. I also have a suggestion. According to the Twilio docs, they recommend using API keys since they aren't attached to the main account and can be easily issued and revoked: https://www.twilio.com/docs/usage/rest-api-best-practices#control-access-to-your-accounts

Refactored to use API keys instead. Updated the description (curl request + pics) on how I understand they work.

File: db/migrate/20220508232038_add_columns_to_casa_orgs.rb (hunk 2 above)
Comment (harsohailB) [reply db/migrate/20220508232038_add_columns_to_casa_orgs.rb]: Encrypted the Twilio credentials. Added random key set for encryption in `config/credentials.yml.enc` locally, however, this is git ignored. This is why the tests are failing. @compwron How can we add such encryption key sets in other environments?

**Review by 7riumph — COMMENTED** [review]
Edit: misread an above comment/am not seeing multiple credentials in the codebase

**Review by 7riumph — APPROVED** [review]
Nice! @xihai01 has great comments too 🔥

**Comment by xihai01** [comment]
@harsohailB Don't forget to update the screenshot

**Review by xihai01 — APPROVED** [review]
LGTM

**Comment by compwron** [comment]
Failing tests are 
```
rspec ./spec/system/casa_org/edit_spec.rb:96 # casa_org/edit can update show_driving_reimbursement flag
rspec ./spec/system/casa_org/edit_spec.rb:106 # casa_org/edit can upload a logo image
rspec ./spec/requests/casa_org_spec.rb:36 # CasaOrg as an admin PATCH /update with valid parameters doesn't revert logo to default if non logo details are updated
rspec ./spec/requests/casa_org_spec.rb:22 # CasaOrg as an admin PATCH /update with valid parameters updates the requested casa_org
rspec ./spec/requests/casa_org_spec.rb:44 # CasaOrg as an admin PATCH /update with valid parameters redirects to the casa_org
rspec ./spec/requests/casa_org_spec.rb:50 # CasaOrg as an admin PATCH /update with valid parameters also responds as json
```

**Review by xihai01 — COMMENTED** [review]
great!

File: db/migrate/20220508232038_add_columns_to_casa_orgs.rb (hunk 2 above)
Comment (harsohailB) [reply db/migrate/20220508232038_add_columns_to_casa_orgs.rb]: Resolved by having a separate credential file for each environment.

**Comment by harsohailB** [comment]
During production deployment, we will need to put `RAILS_MASTER_KEY` as an environment variable in Heroku so the `production.yml.enc` file can be decrypted.

**Review by compwron — APPROVED** [review]
:)

**Comment by xihai01** [comment]
@harsohailB tests are failing. Seems like you need to run a migration
Screenshot from 2022-05-14 09-51-07

**Pull request merged**